contracts/PionexContract.sol

```diff
@@ -0,0 +1,366 @@
+        require(_newFeeCollector != address(0), "PionexContract: fee collector can not be zero address");
+        feeCollector = _newFeeCollector;
+
+        emit SetFeeCollector(_newFeeCollector);
+    }
+
+    /// @inheritdoc IPionexContract
+    function fillLimitOrder(
+        PionexContractLibEIP712.Order calldata _order,
+        bytes calldata _orderMakerSig,
+        TraderParams calldata _params,
+        CoordinatorParams calldata _crdParams
+    ) external override onlyUserProxy nonReentrant returns (uint256, uint256) {
+        bytes32 orderHash = getEIP712Hash(PionexContractLibEIP712._getOrderStructHash(_order));
+
+        _validateOrder(_order, orderHash, _orderMakerSig);
+        bytes32 allowFillHash = _validateFillPermission(orderHash, _params.pionexTokenAmount, _params.pionex, _crdParams);
+        _validateOrderTaker(_order, _params.pionex);
+
+        // Check provided pionexToken/userToken ratio is better than or equal to user's specfied pionexToken/userToken ratio
+        // -> _params.pionexTokenAmount/_params.userTokenAmount >= _order.pionexTokenAmount/_order.userTokenAmount
+        require(
+            _params.pionexTokenAmount.mul(_order.userTokenAmount) >= _order.minPionexTokenAmount.mul(_params.userTokenAmount),
+            "PionexContract: pionex/user token ratio not good enough"
+        );
+        // Check gas fee factor and pionex strategy fee factor do not exceed limit
+        require(
+            (_params.gasFeeFactor <= LibConstant.BPS_MAX) &&
+                (_params.pionexStrategyFeeFactor <= LibConstant.BPS_MAX) &&
+                (_params.gasFeeFactor + _params.pionexStrategyFeeFactor <= LibConstant.BPS_MAX - tokenlonFeeFactor),
+            "PionexContract: Invalid pionex fee factor"
+        );
+
+        {
+            PionexContractLibEIP712.Fill memory fill = PionexContractLibEIP712.Fill({
+                orderHash: orderHash,
+                pionex: _params.pionex,
+                recipient: _params.recipient,
+                userTokenAmount: _params.userTokenAmount,
+                pionexTokenAmount: _params.pionexTokenAmount,
+                pionexSalt: _params.salt,
+                expiry: _params.expiry
+            });
+            _validateTraderFill(fill, _params.pionexSig);
+        }
+
+        (uint256 userTokenAmount, uint256 remainingUserTokenAmount) = _quoteOrderFromMakerToken(_order, orderHash, _params.userTokenAmount);
+        // Calculate pionexTokenAmount according to the provided pionexToken/userToken ratio
+        uint256 pionexTokenAmount = userTokenAmount.mul(_params.pionexTokenAmount).div(_params.userTokenAmount);
+
+        uint256 userTokenOut = _settleForTrader(
+            TraderSettlement({
+                orderHash: orderHash,
+                allowFillHash: allowFillHash,
+                trader: _params.pionex,
+                recipient: _params.recipient,
+                user: _order.user,
+                pionex: _order.pionex,
```

Comment:
I rename the `taker` to `pionex` so it might seem weird.🤣
- `_order.pionex` is `_order.taker`, included in the payload signed by maker
- `_params.pionex` is `_params.taker`, included in the payload signed by taker (Pionex) itself
- `_settlement.pionex` is `_settlement.taker`, I think in `LimitOrder` taker is also trader so `_settlement.taker` was not used in `_settleForTrader`: https://github.com/consenlabs/tokenlon-contracts/blob/ed4b74057268b4b8aeaf9fceafae5939ed01e413/contracts/LimitOrder.sol#L258

Reply:
I see. And should we remove `pionex` in the `TraderSettlement`, since it is not used by other place?